src/Oauth2Client.php

```diff
@@ -0,0 +1,320 @@
+   */
+  public $apiVersion = 'v1';
+
+  /**
+   * Create a new \Shapeways\Oauth2CurlClient
+   *
+   * @param string $clientId your app consumer key
+   * @param string $clientSecret your app consumer secret
+   * @param string|null $redirectUrl your app callback url
+   * @param string|null $accessToken a users oauth token if it is already known
+   * @param string|null $refreshToken if it is already known
+   */
+  public function __construct(
+    $clientId,
+    $clientSecret,
+    $redirectUrl,
+    $accessToken = null,
+    $refreshToken = null
+  ) {
+    $this->clientId = $clientId;
+    $this->clientSecret = $clientSecret;
+    $this->redirectUrl = $redirectUrl;
+    $this->accessToken = $accessToken;
+    $this->refreshToken = $refreshToken;
+  }
+
+  /**
+   * Grant Type 1: Resource owner credentials grant
+   *
+   * https://developers.shapeways.com/quick-start#authenticate
+   *
+   * Your don't need a redirect url for this grant type
+   *
+   * Use "access_token" from result for other API calls
+   *
+   * @return array - json decoded api response
+   */
+  public function generateAccessTokenClientCredentialGrant()
+  {
+    $params = array(
+      "grant_type" => "client_credentials"
+    );
+
+    $url = $this->baseApiUrl . '/oauth2/token';
+
+
+    $process = curl_init($url);
+    curl_setopt($process, CURLOPT_USERPWD, $this->clientId . ":" . $this->clientSecret);
+    curl_setopt($process, CURLOPT_TIMEOUT, 30);
+    curl_setopt($process, CURLOPT_POST, 1);
+    curl_setopt($process, CURLOPT_POSTFIELDS, $params);
+    curl_setopt($process, CURLOPT_RETURNTRANSFER, true);
+    $result = curl_exec($process);
+    curl_close($process);
+
+    return json_decode($result);
+
+  }
+
+
+  /**
+   * Grant Type 2: Authorization code grant
+   * Step 1 - generate a authorization code
+   *
+   *
+   * If you want to use this app for a different Shapeways User account.
+   *
+   *
+   * @link https://developers.shapeways.come/quick-start#authenticate
+   *
+   * redirect_uri - This determines where authentication requests will be sent and received by your app.
+   * Users will be redirected to this URL when they attempt to use your app.
+   */
+  public function generateAccessTokenAuthorizationGrant()
+  {
+    $clientId = $this->clientId;
+    $url = $this->baseApiUrl . "/oauth2/authorize?response_type=code&client_id=" . $clientId . "&redirect_uri=" . rawurlencode($this->redirectUrl);
+    header('Location: ' . $url);
+  }
+
+
+  /**
+   * Grant Type 2: Authorization code grant
+   * Step 2 - Generate access token
+   *
+   * @link https://developers.shapeways.com/quick-start#authenticate
+   *  function for handling call back of generateAccessTokenAuthorizationGrant() call above
+   *
+   * Use "access_token" from result for other API calls
+   *
+   * @return array - json decoded api response
+   */
+  public function handleAuthorizationGrantCallback()
+  {
+    $code = $_REQUEST['code'] ?? null;
```

Comment:
why not let people pass in the code themselves? In case they're storing the code and moving people around, or working with a framework that turns requests into psr7 objects and using that.